File: src/services/OllamaService.js
const OLLAMA_BASE_URL = 'http://localhost:11434'
// Use 'resumeforge' após rodar .\setup-model.ps1 (modelo customizado com system prompt dedicado)
// Por padrão usa llama3.1 que já vem instalado
const MODEL = 'qwen2.5:14b'

/**
 * Verifica se o Ollama está online e se o modelo está disponível.
 * Retorna { ok: true } ou { ok: false, reason: string, models: string[] }
 */
export async function checkModel() {
  try {
    const res = await fetch(`${OLLAMA_BASE_URL}/api/tags`)
    if (!res.ok) {
      return { ok: false, reason: 'ollama_error', models: [] }
    }
    const data = await res.json()
    const models = (data.models || []).map((m) => m.name)
    const available = models.some((name) => name.startsWith(MODEL))
    if (!available) {
      return { ok: false, reason: 'model_missing', models }
    }
    return { ok: true, models }
  } catch (err) {
    if (err instanceof TypeError) {
      // Erro de rede — Ollama offline ou CORS bloqueado
      return { ok: false, reason: 'offline', models: [] }
    }
    return { ok: false, reason: 'unknown', models: [] }
  }
}

/**
 * Envia um prompt ao Ollama com stream: true.
 * Retorna o ReadableStream bruto do response.body.
 * Lança erro se a requisição falhar.
 */
export async function generate(prompt, signal) {
  const res = await fetch(`${OLLAMA_BASE_URL}/api/generate`, {
    method: 'POST',
    signal,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      model: MODEL,
      prompt,
      stream: true,
      format: 'json',
      options: { num_ctx: 8192 },
    }),
  })

  if (!res.ok) {
    throw new Error(`Ollama retornou status ${res.status}`)
  }

  return res.body
}

File: backend/services/ollama_service.py
import httpx, json, re

OLLAMA_BASE = "http://localhost:11434"
MODEL = "qwen2.5:14b"

async def check_model() -> dict:
    try:
        async with httpx.AsyncClient(timeout=5) as client:
            r = await client.get(f"{OLLAMA_BASE}/api/tags")
            r.raise_for_status()
            data = r.json()
            models = [m["name"] for m in data.get("models", [])]
            has_model = any(MODEL in m for m in models)
            if not has_model:
                return {"ok": False, "reason": f"Modelo '{MODEL}' não encontrado. Execute: ollama pull {MODEL}", "models": models}
            return {"ok": True, "reason": None, "models": models}
    except Exception as e:
        return {"ok": False, "reason": f"Ollama não está rodando: {e}", "models": []}

def _strip_fences(text: str) -> str:
    text = text.strip()
    text = re.sub(r'^```(?:json)?\s*', '', text)
    text = re.sub(r'\s*```$', '', text)
    return text.strip()

async def generate_stream(prompt: str) -> dict:
    full_text = ""
    async with httpx.AsyncClient(timeout=300) as client:
        async with client.stream(
            "POST", f"{OLLAMA_BASE}/api/generate",
            json={"model": MODEL, "prompt": prompt, "stream": True,
                  "options": {"num_ctx": 8192, "temperature": 0.3}}
        ) as resp:
            resp.raise_for_status()
            async for line in resp.aiter_lines():
                if not line.strip():
                    continue
                try:
                    chunk = json.loads(line)
                    full_text += chunk.get("response", "")
                    if chunk.get("done"):
                        break
                except Exception:
                    continue

    cleaned = _strip_fences(full_text)
    return json.loads(cleaned)
